### PR TITLE
Optimise instance load/store/delete by skipping special accessors when possible

### PR DESCRIPTION
**Pull request by dpgeorge**
This PR aims to optimise load/store/delete of attributes in user defined classes by not looking up special accessors (property, `__get__`, `__delete__`, `__set__`, `__setattr__` and `__getattr_`) if they are guaranteed not to exist in the class.

Currently, if you do `my_obj.foo()` then the runtime has to do a few checks to see if `foo` is a property or has `__get__`, and if so delegate the call.  And for stores things like `my_obj.foo = 1` has to first check if `foo` is a property or has `__set__` defined on it.

Doing all those checks each and every time the attribute is accessed has a performance penalty.  This PR eliminates all those checks for cases when it's guaranteed that the checks will always fail, ie no attributes are properties nor have any special accessor methods defined on them.

To make this guarantee it checks all attributes of a user-defined class when it is first created.  If any of the attributes of the user class are properties or have special accessors, or any of the base classes of the user class have them, then it sets a flag in the class to indicate that special accessors must be checked for.  Then in the load/store/delete code it checks this flag to see if it can take the short cut and optimise the lookup.

Code size increase with this PR is:
```
   bare-arm:   +16 
minimal x86:   +32 
   unix x64:  +200 
unix nanbox:  +256 
      stm32:   +80 
     cc3200:   +80 
    esp8266:  +436 
      esp32:  +108 
```
Bare-arm and minimal increase because of the introduction of a `flags` entry in `mp_obj_type_t`, which will be generally useful for other things in the future as well.  esp8266 increases by more than the others because it has overhead for loading non-32-bit values from RAM.

With this PR performance increases by about 6%, which is quite an improvement.  More importantly, MICROPY_PY_DESCRIPTORS can now be enabled without any additional overhead (see related discussion in #3644).

Performance tests that were done:
- on unix x86-64, pystone improved by about 5%
- on pyboard, pystone improved by about 6.5%, from 1683 up to 1794
- on pyboard, bm_chaos (from CPython benchmark suite) improved by about 5%
- on esp32, pystone improved by about 30%, but caching effects probably play a role here
- on esp32, bm_chaos improved by about 11%

**One important downside of this PR**: because checks for properties and special accessors is done at the time of creation of a user-defined class, the optimisation trick here will fail if classes have properties or special methods added after they are created.  For example
```python
class A:
    pass
def getter(self):
    print('get')
    return 1
A.foo = property(getter) # dynamically add property to the class
print(A().foo) # won't work with this PR, just returns a property object
```
The reason it doesn't work now is because properties/special accessors are only checked for during class creation.  It would be possible to check when attributes are added to a class if they are a property/special accessor, but that only solves part of the problem: there's still the chance that a *base class* is changed dynamically, and the derived class has no way of knowing this.

In summary: this PR gives a really great performance boost and allows to enable descriptors with no cost to classes that don't use them.  But is it worth giving up on the ability to dynamically add properties/special accessors to a class (does anyone ever do that)?

### CODE REVIEW

**Comment by tannewt** [comment]
Yes! I definitely think its worth it. Properties are much simpler to use than functions for state.

Could you detect the gotcha case and throw an error instead of failing silently later?

Great work! Thank you!

**Comment by dpgeorge** [comment]
> Properties are much simpler to use than functions for state.

But note that performance is not improved for classes with properties, it will be the same as before.  It's just improved for user-defined classes that don't use properties or any special accessors (of which there are a lot).

> Could you detect the gotcha case and throw an error instead of failing silently later?

It can be detected at class creation time: if the new class has any base classes at all then the optimisation can be disabled.  But this would reduce the effectiveness of the performance boost: only classes with no base classes and no properties would have improved efficiency.  [It would be possible to further optimise classes that have only a single parent, as long as that parent doesn't have any parents.]

**Comment by stinos** [comment]
Yes looks worth it, but silent fail could cause major headaches. I hardly ever use properties at all myself, let alone dynamically created ones, but sometimes you port some code and then it might go unnoticed. Would it be an option to detect things like `A.foo = property(getter)` and assignment of special methods after class creation and just always raise an exception (to also cover bas class case)? Still safer.

File: py/obj.h [hunk 1]
```diff
@@ -465,8 +465,12 @@ struct _mp_obj_type_t {
...
     mp_obj_base_t base;
 
-    // The name of this type.
-    qstr name;
+    // Flags associated with this type.
+    #define MP_OBJ_TYPE_FLAG_xyz (0x0001)
+    uint16_t flags;
+
+    // The name of this type, a qstr.
+    uint16_t name;
```
Comment (stinos) [inline py/obj.h]: Just to doublecheck: a qstr is a size_t but it's ok to cast it to uint16_t right?

**Comment by nickovs** [comment]
This is great!

As for the downside regarding changing the class definition after creation, this is a shame but I suspect it is rarely a problem.

It looks like with the current implementation this is only an issue if the _first descriptor_ is added to a class after creation, since if there are any descriptors then the flag is already set and the search will go ahead. This makes the behaviour even less of a problem, although it does potentially make the edge case somewhat stranger and so more likely to catch someone out. If uPy ever supports proper metaclasses then we will probably want to make sure that the flag gets set at the right place in the __new__ process.

**Comment by dpgeorge** [comment]
> but silent fail could cause major headaches.

Yes I agree, I don't like the idea of silently doing the wrong thing.  

> Would it be an option to detect things like A.foo = property(getter) and assignment of special methods after class creation and just always raise an exception

Yes that would be pretty easy to check, there is only one location in the code where the check needs to be added, and it's not a hot spot so won't degrade performance.

**Comment by dpgeorge** [comment]
> It looks like with the current implementation this is only an issue if the first descriptor is added to a class after creation, since if there are any descriptors then the flag is already set and the search will go ahead.

Yes that's right.

It would be possible to check if descriptors are added after creation of a class, and set the flag then (ie dynamically update the flag when new methods are added after creation).  Then the only case that is not caught is if a descriptor is added to a class after creation *and* that class has already been subclassed (since the subclass won't know about the new descriptor).

That could be fixed by, at each access of an attribute, the flag for special accessors is checked in the current class and its immediate parent.  This pushes the unchecked cases to when a descriptor is added to a class after creation and that class has already been subclassed by 2 generation (eg A->B->C and A is updated after C is defined).

How about this for a solution: there is a second flag which indicates if a class has been subclassed or not.  If it has not then it allows properties/descriptors to be added dynamically after class creation.  As soon as it is subclassed it disallows adding them (unless at least one has already been added).  So basically as soon as you subclass something you "lock in" the fact that it does or does not have properties/descriptors.  This scheme would allow pretty good flexibility and CPython compatibility, and raise an exception if you step outside the supported behaviour.

**Comment by dpgeorge** [comment]
> If the restriction is that you can not go from zero descriptors to one descriptor on a class after that class has been subclassed, we raise an exception when someone tries, and all other cases work,

Ok, I've implemented this behaviour now in this PR.

**Comment by nickovs** [comment]
Maybe I'm missing something (or maybe not) but it looks like the code [here](https://github.com/micropython/micropython/pull/3826/files#diff-9b89b5d4c6e904e03d6459f4075a4770R1040) sets the attribute and the flag and only then raises the exception, rather than stopping the attribute from being set. Maybe this intentional (i.e. do what the user asked but moan about it) but I just wanted to check.

**Comment by dpgeorge** [comment]
> Maybe I'm missing something (or maybe not) but it looks like the code here sets the attribute and the flag and only then raises the exception, rather than stopping the attribute from being set.

Yes you are right, it does let the attribute in the class be set, and then does a post-check.

> Maybe this intentional (i.e. do what the user asked but moan about it) but I just wanted to check.

I didn't think too hard about this point... I just thought that a RuntimeError is enough for the user to realise there's a major failure with the system and they shouldn't try to continue.

But probably it is better to do a pre-check.

**Comment by nickovs** [comment]
That said, there's an argument for things that fail not having too many side effects.

**Comment by dpgeorge** [comment]
I updated this PR to do a pre-check for class stores, so it doesn't have a side effect if the store fails.

**Comment by stinos** [comment]
Looks good!

**Comment by nickovs** [comment]
Looks good to me.

Will you enable descriptors by default when you merge this in? Other than the memory impact on esp8266 it looks like there's not much down side to doing so now.

**Comment by dpgeorge** [comment]
> Will you enable descriptors by default when you merge this in?

Yes I would like to enable it on stm32, esp8266 and esp32 (it's already enabled on unix).

**Comment by dpgeorge** [comment]
Just a minor note: when trying to add a new property/descriptor to an already-subclassed class which doesn't already have any special accessors, I changed the exception raised from RuntimeError to AttributeError.  I feel AttributeError better describes the problem, because that's what you'd get if you try to assign an attribute to something that can't have it assigned (eg because it's a built-in class).

**Comment by dpgeorge** [comment]
Finally, let me argue that this PR is an overall good addition: it's purely a code optimisation and costs bytes (trading text bytes for speed), but is it worth spending the bytes on this particular optimisation?

The cost in code space bytes for the optimisation is:
```
   unix x64:  +528 
unix nanbox:  +508 
      stm32:  +192 
     cc3200:  +200 
    esp8266:  +332 
      esp32:  +244
```

On pyboard it's an increase of 0.06% in code size for a gain (in pystone performance) of roughly 6.5%.

Compare that with compiling pyboard code with -O2 optimisation (vs -Os), which increases code size by about 10% for a gain in (pystone) performance of about 10%.

So in relative terms it's quite a lot of perfermance for very little cost.

Furthermore, it's an optimisation that's pretty widely applicable because it improves lookup performance for all methods of user defined classes, and stores of attributes, at least for those that don't have special accessors.

And, it allows to enable descriptors with minimal additional runtime overhead if they are not used.

File: py/objtype.c [hunk 2]
```diff
@@ -591,6 +597,11 @@ STATIC void mp_obj_instance_load_attr(mp_obj_t self_in, qstr attr, mp_obj_t *des
     mp_obj_class_lookup(&lookup, self->base.type);
     mp_obj_t member = dest[0];
     if (member != MP_OBJ_NULL) {
+        if (!(self->base.type->flags & TYPE_FLAG_HAS_SPECIAL_ACCESSORS)) {
```
Comment (pfalcon) [inline py/objtype.c]: Why this overhead code isn't wrapped with `#if ENABLE_SPECIAL_ACCESSORS`?

File: py/objtype.c [hunk 3]
```diff
@@ -652,11 +663,15 @@ STATIC void mp_obj_instance_load_attr(mp_obj_t self_in, qstr attr, mp_obj_t *des
 STATIC bool mp_obj_instance_store_attr(mp_obj_t self_in, qstr attr, mp_obj_t value) {
     mp_obj_instance_t *self = MP_OBJ_TO_PTR(self_in);
 
+    if (!(self->base.type->flags & TYPE_FLAG_HAS_SPECIAL_ACCESSORS)) {
```
Comment (pfalcon) [inline py/objtype.c]: Why this overhead code isn't wrapped with `#if ENABLE_SPECIAL_ACCESSORS`?

File: py/objtype.c (hunk 2 above)
Comment (dpgeorge) [reply py/objtype.c]: I checked this (wrapping it in `ENABLE_SPECIAL_ACCESSORS`) and it doesn't do anything, it seems the compiler is smart enough to eliminate it anyway.  That's because if there are no special accessors none of the code below exists, and the statement after this `if` is immediately a `return`.

File: py/objtype.c (hunk 3 above)
Comment (dpgeorge) [reply py/objtype.c]: Same as above, the compiler is smart enough to eliminate it.  And wrapping this with `ENABLE_SPECIAL_ACCESSORS` then requires wrapping the label `skip_special_accessors` to avoid compiler warnings, which makes the code more ugly / harder to read.

**Comment by dpgeorge** [comment]
This was merged in 36c105218321167d47b0fb67575a7cddc36d17e7.  Thanks for all the feedback and suggestions to improve it.

**Comment by nickovs** [comment]
> Thanks for all the feedback and suggestions to improve it.

@dpgeorge Thanks for building this!
